packages/rpc-core/src/rpc-methods/getProgramAccounts.ts

```diff
@@ -0,0 +1,185 @@
+    lamports: LamportsUnsafeBeyond2Pow53Minus1;
+    /** pubkey of the program this account has been assigned to */
+    owner: Base58EncodedAddress;
+    /** the epoch at which this account will next owe rent */
+    rentEpoch: U64UnsafeBeyond2Pow53Minus1;
+    /** the data size of the account */
+    space: U64UnsafeBeyond2Pow53Minus1;
+}>;
+
+type AccountInfoWithDefaultData = Readonly<{
+    data: Base58EncodedBytes;
+}>;
+
+type AccountInfoWithBase58EncodedData_DEPRECATED = Readonly<{
+    data: Base58EncodedDataResponse;
+}>;
+
+type AccountInfoWithBase64EncodedData = Readonly<{
+    data: Base64EncodedDataResponse;
+}>;
+
+type AccountInfoWithBase64EncodedZStdCompressedData = Readonly<{
+    data: Base64EncodedZStdCompressedDataResponse;
+}>;
+
+type AccountInfoWithJsonData = Readonly<{
+    data:
+        | Readonly<{
+              /** Name of the program that owns this account. */
+              program: string;
+              parsed: unknown;
+              space: U64UnsafeBeyond2Pow53Minus1;
+          }>
+        // If `jsonParsed` encoding is requested but a parser cannot be found for the given
+        // account the `data` field falls back to `base64`.
+        | Base64EncodedDataResponse;
+}>;
+
+type AccountWithPubkey<TAccount extends AccountInfoBase> = Readonly<{
+    account: TAccount;
+    pubkey: Base58EncodedAddress;
+}>;
+
+type GetProgramAccountsMemcmpFilter = Readonly<{
+    offset: U64UnsafeBeyond2Pow53Minus1;
+    bytes: string;
+    encoding: 'base58' | 'base64';
+}>;
+
+type GetProgramAccountsDatasizeFilter = Readonly<{
+    dataSize: U64UnsafeBeyond2Pow53Minus1;
+}>;
+
+type GetProgramAccountsApiCommonConfig = Readonly<{
+    /** @default "finalized" */
+    commitment?: Commitment;
+    /** The minimum slot that the request can be evaluated at */
+    minContextSlot?: Slot;
+    /** filter results (up to 4 filters allowed) @see https://docs.solana.com/api/http#filter-criteria */
+    filters?: (GetProgramAccountsMemcmpFilter | GetProgramAccountsDatasizeFilter)[];
+}>;
+
+type GetProgramAccountsApiSliceableCommonConfig = Readonly<{
+    /** Limit the returned account data */
+    dataSlice?: DataSlice;
+}>;
+export interface GetProgramAccountsApi {
+    /**
+     * Returns the account information for a list of Pubkeys.
+     */
+    getProgramAccounts(
+        program: Base58EncodedAddress,
+        config: GetProgramAccountsApiCommonConfig &
+            GetProgramAccountsApiSliceableCommonConfig &
+            Readonly<{
+                encoding: 'base64';
+                withContext: true;
+            }>
+    ): RpcResponse<AccountWithPubkey<AccountInfoBase & AccountInfoWithBase64EncodedData>[]>;
+
+    getProgramAccounts(
+        program: Base58EncodedAddress,
+        config: GetProgramAccountsApiCommonConfig &
+            GetProgramAccountsApiSliceableCommonConfig &
+            Readonly<{
+                encoding: 'base64';
+                withContext?: boolean;
```

Comment:
This looks super weird, but it works apparently? If provided as `true` it's going to the first overload instead, so this covers `undefined` or `false` yeah?

Reply:
I also only recently discovered you can do [default parameters in TypeScript](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAcwKZQAoCc7KwQwFsBBCCOcKAZwAoAoRRxfAExa1SqoC5EqosMMMgA0iBk3JhgMZLwDeEpk1SQ4LIcgD8vfoOEilyxAHcYUABYBhBFFQAPKDsQAjOHAA2qfGEPKAvogAvIiKxipqGsK8AOQu+FSoAKwAHDF+4abm1rYOULzA+B6JGYz+dACUoUZSVJ6oAHQeuDSs7JxUFQDcNQh1Xk0tUjLI3XT+QA), like Python:

Reply:
Yep exactly, it's basically that the param is optional and the default is `false`. So you define `withContext: true` with its `RpcResponse<...>` and then `withContext?: boolean` will capture both explicit false and undefined -> default false.
Annoyingly it doesn't work with `encoding` though because (for this RPC method...) no defined encoding has its own annoying special case
I don't think default parameters will solve it here either because you want a different type on the response depending on the value of that parameter - it's not just input. So I think that requires an overload, or the conditional typing that I couldn't get to work

Reply:
Sounds good to me. You're right that is a mega-pain lol